src/engine/CodeValidator.v3

```diff
@@ -1351,10 +1337,20 @@ class CodeValidator(extensions: Extension.set, limits: Limits, module: Module, e
 			if (!ValueTypes.isAssignable(ct[j], p[j])) err_atpc().TypeMismatchInHandler("catch", i, p[i], ct[i]);
 		}
 	}
-	def checkContHandle(i: int, tag: TagDecl, target: ControlEntry) {
+	def readAndCheckContHandlerTable(cont: ContDecl) {
+		var handlers = parser.readContHandlers();
+		for (i < handlers.length) {
+			var handler = handlers[i];
+			var target = getControl(handler.depth);
+			if (target == null) return;
+			checkContHandle(i, handler.tag, cont, target);
+			ctlxfer.refC(target, handler.tag, false);
+		}
+	}
+	def checkContHandle(i: int, tag: TagDecl, cont: ContDecl, target: ControlEntry) {
 		var p = labelArgs(target);
 		if (p == null) return;
-		var ct = Arrays.append(ValueType.Ref(true, HeapType.Cont(null)), tag.fields);
+		var ct = Arrays.append(ValueType.Ref(true, HeapType.Cont(cont)), tag.fields);
```

Comment:
My reading of the rule for `resume` is that the parameter types of the continuation expected at the label are the results from the tag, and the result types of the continuation expected at the label are the same (or subtypes) of the result of the original `cont`.